src/aleph/web/controllers/storage.py

```diff
@@ -56,20 +80,133 @@ async def add_storage_json_controller(request: web.Request):
+        None, functools.partial(encode_defunct, text=verification.decode("utf-8"))
+    )
+
+    verified = False
+    try:
+        # we assume the signature is a valid string
+        address = await run_in_executor(
+            None,
+            functools.partial(
+                Account.recover_message, message_hash, signature=message.signature
+            ),
+        )
+        if address == message.sender:
+            verified = True
+        else:
+            return False
+
+    except Exception as e:
+        verified = False
+    return verified
+
+
+async def get_message_content(
+    post_data: MultiDictProxy[Union[str, bytes, FileField]]
+) -> Tuple[dict, int]:
+    message_bytearray = post_data.get("message", b"")
+    value = post_data.get("size") or 0
+
+    if isinstance(message_bytearray, bytearray):
+        message_string = message_bytearray.decode("utf-8")
+        message_dict = json.loads(message_string)
+        message_dict["time"] = float(message_dict["time"])
+    else:
+        message_dict = {}
+
+    return message_dict, int(str(value))
+
+
+async def verify_and_handle_request(
+    pending_message_db, file_io, message, size, session_factory
+):
+    content = file_io.read(size)
+    item_content = json.loads(message["item_content"])
+    actual_item_hash = sha256(content).hexdigest()
+    c_item_hash = item_content["item_hash"]
+
+    is_signature = await verify_signature(message=pending_message_db)
+    with session_factory() as session:
+        current_balance = get_total_balance(
+            session=session, address=pending_message_db.sender
+        )
+    if current_balance < len(content):
+        output = {"status": "Payment Required"}
+        return web.json_response(output, status=402)
+    if not is_signature:
+        output = {"status": "Forbidden"}
+        return web.json_response(output, status=403)
+    elif actual_item_hash != c_item_hash:
+        output = {"status": "Unprocessable Content"}
+        return web.json_response(output, status=422)
+    elif len(content) > 25_000 and not message:
+        output = {"status": "Unauthorized"}
+        return web.json_response(output, status=401)
+    else:
+        return None
+
+
+async def storage_add_file_with_message(request: web.Request):
     storage_service = get_storage_service_from_request(request)
     session_factory = get_session_factory_from_request(request)
+    config = get_config_from_request(request)
 
-    # No need to pin it here anymore.
-    # TODO: find a way to specify linked ipfs hashes in posts/aggr.
     post = await request.post()
     file_io = multidict_proxy_to_io(post)
+    message, size = await get_message_content(post)
+    pending_message_db = PendingMessageDb.from_message_dict(
+        message_dict=message, reception_time=dt.datetime.now(), fetched=True
+    )
+    mq_channel = await get_mq_channel_from_request(request, logger=logger)
+    mq_queue = await mq_make_aleph_message_topic_queue(
```

Comment:
The MQ stuff only needs to be done if a) there is a message attached to the request and b) the user specified `sync=True` in the request. The `sync` parameter is nowhere to be found in your PR, add it.